### ISSUE TITLE
Cryst 4.1.12: ImagesRepresentative can return nonsense results

### ISSUE THREAD

**Issue by dominicelse** [gap-system/gap#913]
Running the code:

```
G := SpaceGroup(3,17);
iso := IsomorphismPcpGroup(G);
A := [ [ -1, 0, 0, 0 ], [0, -1, 0, 0], [0, 0, 1, 0], [1/2, 0, 0, 1] ];
A in G;
ImagesRepresentative(iso, A);
```
### Observed behaviour

```
true
g2*g3^1/2
```
### Expected behaviour

The output should be a product of generators with _integer_ exponents. Looking at the code in `pcpgrp.gi`, the logic of the function seems to be assuming a symmorphic space group.
### Copy and paste GAP banner (to tell us about your setup)

```
 ┌───────┐   GAP 4.8.4, 04-Jun-2016, build of 2016-08-09 17:10:16 (PDT)
 │  GAP  │   http://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-gcc-default64
```

**Comment by alex-konovalov** [comment]
I can reproduce this with GAP 4.8.5. Some of the calculations require Alnuth package to be usable, and that requires Pari/GP to be installed, so I've actually used GAP Docker container from https://hub.docker.com/r/gapsystem/gap-docker/ as below:
```
docker pull gapsystem/gap-docker
docker run --rm -i -t gapsystem/gap-docker gap
```

First, `SpaceGroup` is the function from the AcLib package from `gap4r8/pkg/aclib/gap/crystgrp.gi` (from line 1285) but it is not documented. The Cryst package provides a family of documented `SpaceGroup...` functions. What space group does is that it calls `SpaceGroupBBNWZ` from CrystCat with appropriate arguments.
 
Next, `ApplicableMethod(IsomorphismPcpGroup,[G])` points to the method in`gap4r8/pkg/cryst/gap/pcpgrp.gi` (from line 40) for `AffineCrystGroupOnLeftOrRight` and `ApplicableMethod(ImagesRepresentative,[iso,A])` - to the method in `gap4r8/pkg/cryst/gap/pcpgrp.gi` (from line 147) for phi in `IsFromAffineCrystGroupToPcpGroup`.

**Comment by alex-konovalov** [comment]
Next, I've "erased" the additional knowledge about the group, and I do not have fractional exponents any more:
```
gap> G := SpaceGroup(3,17);
SpaceGroupOnRightBBNWZ( 3, 3, 1, 1, 2 )
gap> H := Group(GeneratorsOfGroup(G));
<matrix group with 5 generators>
gap> iso := IsomorphismPcpGroup(H);
[ [ [ -1, 0, 0, 0 ], [ 0, -1, 0, 0 ], [ 0, 0, 1, 0 ], [ -1/2, 1, 0, 1 ] ], 
  [ [ -1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, -1, 0 ], [ 0, 0, 1, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ 0, 0, 1, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ -1, 1, 0, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ 1, 0, 0, 1 ] ] ] -> [ g1, g2, g3, g4, g5 ]
gap> A := [ [ -1, 0, 0, 0 ], [0, -1, 0, 0], [0, 0, 1, 0], [1/2, 0, 0, 1] ];
[ [ -1, 0, 0, 0 ], [ 0, -1, 0, 0 ], [ 0, 0, 1, 0 ], [ 1/2, 0, 0, 1 ] ]
gap> A in H;
true
gap> ImagesRepresentative(iso, A);
g1*g4^-1
```

**Comment by alex-konovalov** [comment]
@dominicelse thank you for reporting this - I have informed the authors of the above mentioned packages and asked for their help.

**Comment by fingolfin** [comment]
I cannot reproduce this, with GAP development version, CrystCat 1.1.6, Aclib 1.2, Cryst 4.1.12, Alnuth 3.0.0. Instead I get this:
```
gap> Print(ImagesRepresentative(iso, A));
g1
gap> H:=Image(iso);
Pcp-group with orders [ 2, 2, 0, 0, 0 ]
gap> PreImage(iso,H.1) = A;
true
```

Which versions of these packages do you two use?

**Comment by alex-konovalov** [comment]
Same versions as you, @fingolfin

**Comment by fingolfin** [comment]
OK, if I start gap with -A, and only load aclib, it works. If I also load polenta, it breaks down. So either a bug in polenta, or a bad interaction between it and the other packages, is to be blamed here.

**Comment by fingolfin** [comment]
The difference is introduced in this line of the `ImagesRepresentative` method:
```
isoP := IsomorphismPcpGroup( PointGroup( S ) );
```
Different methods are invoked depending on whether polenta is loaded or not. Unfortunately, the point group is both in the filter `IsPointGroup` and `IsRationalMatrixGroup`, which have exactly equal rank:
```
gap> RankFilter(IsPointGroup);
50
gap> RankFilter(IsRationalMatrixGroup);
50
```
Note that both methods return valid isomorphisms, just into different groups. The result is that one maps our group element to exponent vector [0,1], the other to exp vector [1,0].

One immediate fix would be to increase the rank of the `IsPointGroup` method (or increase the rank of `IsPointGroup`). But I think this just covers the real bug: That `cryst` is implicitly expecting that *this precise way of constructing an isomorphism* is used. That seems very dangerous. 
At the very least, this should be made explicit, and tested for somehow.

**Comment by fingolfin** [comment]
OK, one step closer: Part of the problem is that in the construction of the iso for the spacegroup, the code does this:
```
    P   := PointGroup( S );
    N   := NiceObject( P );
    ...
    iso := IsomorphismPcpGroup( N );
```
so in short: `IsomorphismPcpGroup( NiceObject( PointGroup( S ) ) )`
But then later, `ImagesRepresentative` assumes that this is equal to `IsomorphismPcpGroup( PointGroup( S ) )`.

So a proper fix would be to to make sure that both methods use the same sequence to obtain the iso. The easier way is to adjust `ImagesRepresentative` to go via the NiceObject. E.g. replace
```
isoP := IsomorphismPcpGroup( PointGroup( S ) );
```
by this:
```
P := PointGroup( S );
isoP := NiceMonomorphism( P ) * IsomorphismPcpGroup( NiceObject( P ) );
```

**Comment by alex-konovalov** [comment]
Thanks @fingolfin - just to check my understanding, am I right that the fix requires *only* package updates, but no changes in the kernel? If this is true, then *which* packages?

**Comment by fingolfin** [comment]
@alex-konovalov Thanks for the question, and sorry for not being that clear on the details in what I wrote before. I'll try to clarify:


This is a bug in package `cryst`, and any fix should be made there (though if the `cryst` authors are unresponsive, I could also make a change to `polenta` to workaround the issue -- but again, the actual cause for the bug is in `cryst`).

To fix the problem, the `cryst` authors should modify `gap/pcpgrp.gi`. There, they should apply this diff:
```patch
diff --git a/cryst/gap/pcpgrp.gi b/cryst/gap/pcpgrp.gi
index b0d7577..8e84903 100644
--- a/cryst/gap/pcpgrp.gi
+++ b/cryst/gap/pcpgrp.gi
@@ -151,14 +151,15 @@ InstallMethod( ImagesRepresentative, FamSourceEqFamElm,
      IsMultiplicativeElementWithInverse ], 0,
 function( iso, elm )
 
-    local d, S, T, elmP, isoP, exp, genS, elm2, v;
+    local d, S, T, elmP, P, isoP, exp, genS, elm2, v;
 
     d := Length( elm ) - 1;
     S := Source( iso );
     T := TranslationBasis( S );
 
     elmP := elm{[1..d]}{[1..d]};
-    isoP := IsomorphismPcpGroup( PointGroup( S ) );
+    P := PointGroup( S );
+    isoP := NiceMonomorphism( P ) * IsomorphismPcpGroup( NiceObject( P ) );
     exp  := Exponents( ImagesRepresentative( isoP, elmP ) );
 
     genS := MappingGeneratorsImages( iso )[1];
```

**Comment by alex-konovalov** [comment]
Unfortunately, no release of `cryst` happened yet, so the problem is still there in GAP 4.8.6 distribution.

**Comment by gaehler** [comment]
Sorry for the delay - I finally published version 4.1.13 containing the proposed fix. Unfortunately, the fix just didn't make it into GAP 4.8.9. But there is hope for GAP 4.9.

**Comment by alex-konovalov** [comment]
Cryst 4.1.13 included in GAP 4.8.10 distribution. Now there I have:  
```
gap> G := SpaceGroup(3,17);
SpaceGroupOnRightBBNWZ( 3, 3, 1, 1, 2 )
gap> iso := IsomorphismPcpGroup(G);
[ [ [ -1, 0, 0, 0 ], [ 0, -1, 0, 0 ], [ 0, 0, 1, 0 ], [ 1/2, 0, 0, 1 ] ], 
  [ [ -1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, -1, 0 ], [ 0, 0, 0, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ 1, 0, 0, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ 0, 1, 0, 1 ] ], 
  [ [ 1, 0, 0, 0 ], [ 0, 1, 0, 0 ], [ 0, 0, 1, 0 ], [ 0, 0, 1, 1 ] ] ] -> 
[ g1, g2, g3, g4, g5 ]
gap> A := [ [ -1, 0, 0, 0 ], [0, -1, 0, 0], [0, 0, 1, 0], [1/2, 0, 0, 1] ];
[ [ -1, 0, 0, 0 ], [ 0, -1, 0, 0 ], [ 0, 0, 1, 0 ], [ 1/2, 0, 0, 1 ] ]
gap> A in G;
true
gap> ImagesRepresentative(iso, A);
g1
gap> H:=Image(iso);
Pcp-group with orders [ 2, 2, 0, 0, 0 ]
gap> PreImage(iso,H.1) = A;
true
```
